File: LLD/Design-Patterns/Builder/code/Builder.java
// Builder Pattern — assembling a Computer with optional parts.
// Compile: javac Builder.java
// Run:     java Builder

class Computer {
    private final String cpu;
    private final int ramGB;
    private final int storageGB;
    private final boolean hasGPU;

    private Computer(ComputerBuilder builder) {
        this.cpu = builder.cpu;
        this.ramGB = builder.ramGB;
        this.storageGB = builder.storageGB;
        this.hasGPU = builder.hasGPU;
    }

    public String toString() {
        return "Computer{cpu=" + cpu + ", ramGB=" + ramGB +
               ", storageGB=" + storageGB + ", hasGPU=" + hasGPU + "}";
    }

    static class ComputerBuilder {
        private String cpu;
        private int ramGB;
        private int storageGB;
        private boolean hasGPU = false;

        ComputerBuilder setCPU(String cpu) {
            this.cpu = cpu;
            return this;
        }

        ComputerBuilder setRAM(int ramGB) {
            this.ramGB = ramGB;
            return this;
        }

        ComputerBuilder setStorage(int storageGB) {
            this.storageGB = storageGB;
            return this;
        }

        ComputerBuilder addGPU() {
            this.hasGPU = true;
            return this;
        }

        Computer build() {
            return new Computer(this);
        }
    }
}

public class Builder {
    public static void main(String[] args) {
        Computer officePC = new Computer.ComputerBuilder()
                .setCPU("Intel i5")
                .setRAM(16)
                .setStorage(512)
                .build();

        Computer gamingPC = new Computer.ComputerBuilder()
                .setCPU("Intel i9")
                .setRAM(32)
                .setStorage(2000)
                .addGPU()
                .build();

        System.out.println(officePC);
        System.out.println(gamingPC);
    }
}

File: v1/LLD/Design-Patterns/Builder/code/builder.js
/**
 * Builder Pattern — assembling a Computer with optional parts.
 * Run: node builder.js
 */

class Computer {
  constructor({ cpu, ramGB, storageGB, hasGPU }) {
    this.cpu = cpu;
    this.ramGB = ramGB;
    this.storageGB = storageGB;
    this.hasGPU = hasGPU;
  }

  toString() {
    return `Computer{cpu=${this.cpu}, ramGB=${this.ramGB}, storageGB=${this.storageGB}, hasGPU=${this.hasGPU}}`;
  }
}

class ComputerBuilder {
  constructor() {
    this._cpu = null;
    this._ramGB = 0;
    this._storageGB = 0;
    this._hasGPU = false;
  }

  setCPU(cpu) {
    this._cpu = cpu;
    return this;
  }

  setRAM(ramGB) {
    this._ramGB = ramGB;
    return this;
  }

  setStorage(storageGB) {
    this._storageGB = storageGB;
    return this;
  }

  addGPU() {
    this._hasGPU = true;
    return this;
  }

  build() {
    return new Computer({
      cpu: this._cpu,
      ramGB: this._ramGB,
      storageGB: this._storageGB,
      hasGPU: this._hasGPU,
    });
  }
}

const officePC = new ComputerBuilder()
  .setCPU("Intel i5")
  .setRAM(16)
  .setStorage(512)
  .build();

const gamingPC = new ComputerBuilder()
  .setCPU("Intel i9")
  .setRAM(32)
  .setStorage(2000)
  .addGPU()
  .build();

console.log(officePC.toString());
console.log(gamingPC.toString());

module.exports = { Computer, ComputerBuilder };
